---
title: "TMB TCGA analysis"
output:
  html_document:
    df_print: paged
---

```{r}
# TMB TCGA check
library(tidyverse)
library(bedr)
#library(rtracklayer)
library(splitstackshape)
library(gridExtra)
library(scales)
library(R.utils)
library(caret)
library(cutpointr)
#setwd("/mnt/hgfs/danhovelson/Box Sync/Strata/")
```

```{r load tcga file}
# keep coding variation only
#tcga <- tcgaraw[!tcgaraw$Variant_Classification %in% #c("Intron","3'Flank","3'UTR","5'Flank","5'UTR","RNA","Splice_Site","Translation_Start_Site"),]
#save(tcga,file="../../analysis/tmb/20180525.tcga_pancx_coding_variants.rda")
#load("/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.rda")

# Keep subset of columns to reduce memory impact

#tcgatrim = tcga[,c("Chromosome","Start_Position","End_Position","Hugo_Symbol","Reference_Allele","Tumor_Seq_Allele1","Tumor_Seq_Allele2","Variant_Type","Variant_Classification","Consequence","Feature_type","VARIANT_CLASS","Tumor_Sample_Barcode","t_depth","t_ref_count","t_alt_count")]
#save(tcgatrim,file="/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.trim_cols.rda")

# only re-load if needed
#if(FALSE) {
  load(file="/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.trim_cols.rda")
#}
```

```{r process tcga data}
tcga_ = tcgatrim

# set variant_id variable for full tcga coding variant index
tcgavid <- read_tsv("/mnt/hgfs/danhovelson/analysis/bed/20180525.tcga_pancx_coding_variants.pos_idx.chr.bed",col_names = FALSE)
colnames(tcgavid) <- c("Chromosome","Start_Position","End_Position","variant_id")
tcgavid %>%
  mutate(variant_hash = paste(gsub("chr","",Chromosome),Start_Position,End_Position,sep=";")) %>%
  select(variant_hash, variant_id) -> tcgavid2
remove(tcgavid)

# match index variable to detailed tcga calls df
tcga <- tcga_ %>%
  mutate(variant_hash = paste(Chromosome,Start_Position,End_Position,sep=";")) %>%
  left_join(tcgavid2,by=c("variant_hash"))
remove(tcga_)

# calculate variant_fraction & create categories/thresholds
tcga$VF <- tcga$t_alt_count / tcga$t_depth
tcga$VFcat <- cut(tcga$VF,c(seq(0,1.0,0.25)),include.lowest=TRUE)
tcga$VFthresh1 <- cut(tcga$VF,c(0,0.1,1.0),include.lowest=TRUE)
tcga$VFthresh2 <- cut(tcga$VF,c(0,0.15,1.0),include.lowest=TRUE)
```
```{r TCGA VF by depth, eval=FALSE}
# load total tcga df
#load("/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.trim_cols.rda")

# calculate tcga vf and depth categories
tcgatrim$vf <- tcgatrim$t_alt_count/tcgatrim$t_depth
tcgatrim$vfcat <- cut(tcgatrim$vf,c(seq(0,0.5,0.1),1.0),include.lowest=TRUE)
tcgatrim$depthcat <- cut(tcgatrim$t_depth,c(0,20,30,40,50,100,1000,1000000),include.lowest=TRUE)

# check vf vs. depth relationship
ggplot(tcgatrim[tcgatrim$Chromosome %in% c(1),],aes(vf,t_depth)) + geom_point(alpha=0.1) + scale_y_log10()
ggplot(tcgatrim,aes(vfcat)) + geom_bar(aes(fill=depthcat))

# read in tcga truth mut rate (WES-based)
tcgatruth <- read_tsv("/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.wes_muts.gt_vf0.rate_tbl.txt")

# check effects of vf/coverage on mut rate calculation
# ex: vf > 20%
tcgafilter <- tcgatrim %>% filter(vf > 0.2)
tfiltmut <- as.data.frame(table(tcgafilter$Tumor_Sample_Barcode))
colnames(tfiltmut) <- c("Tumor_Sample_Barcode","filt_mut_count")
tfiltmut$filt_mut_mb <- tfiltmut$filt_mut_count/30
tcgatruth %>% left_join(tfiltmut,by=c("Tumor_Sample_Barcode")) %>% ggplot(aes(filt_mut_mb,wes_mut_mb)) + geom_point() + ggtitle("VF > 20%")

# ex: coverage > 50
tcgafilter <- tcgatrim %>% filter(t_depth > 50)
tfiltmut <- as.data.frame(table(tcgafilter$Tumor_Sample_Barcode))
colnames(tfiltmut) <- c("Tumor_Sample_Barcode","filt_mut_count")
tfiltmut$filt_mut_mb <- tfiltmut$filt_mut_count/30
tcgatruth %>% left_join(tfiltmut,by=c("Tumor_Sample_Barcode")) %>% ggplot(aes(filt_mut_mb,wes_mut_mb)) + geom_point() + ggtitle("TotalDepth > 50")
```

```{r TCGA calls in TML regions}
# ingest tcga-in-TML index, update to correct variant_id
tcga_tml_idx_ <- read_tsv("/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.IN_TML.uniq.bed",col_names = FALSE)
colnames(tcga_tml_idx_) <- c("Chromosome","Start_Position","End_Position","old_variant_id")
tcga_tml_idx_ %>%
  mutate(variant_hash = paste(gsub("chr","",Chromosome),Start_Position,End_Position,sep=";")) %>%
  left_join(tcgavid2,by=c("variant_hash")) %>%
  select(-old_variant_id,-variant_hash) -> tcga_tml_idx

# filter TCGA calls to those in all TML regions
tcga_tml <- tcga %>%
  filter(variant_id %in% unique(tcga_tml_idx$variant_id)) 

## check pool 1 vs pool 2 vs all
tcga_tml_pool1_idx <- read_tsv("/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.IN_TML_pool1.bed",col_names = FALSE)
colnames(tcga_tml_pool1_idx) <- c("Chromosome","Start_Position","End_Position","j1")
tcga_tml_pool1_idx$pool1 <- "yes"
tcga_tml_pool1_idx %>%
  mutate(variant_hash = paste(gsub("chr","",Chromosome),Start_Position,End_Position,sep=";")) %>%
  select(variant_hash,pool1) %>%
  unique() -> ttmlpool1

tcga_tml_pool2_idx <- read_tsv("/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.IN_TML_pool2.bed",col_names = FALSE)
colnames(tcga_tml_pool2_idx) <- c("Chromosome","Start_Position","End_Position","j1")
tcga_tml_pool2_idx$pool2 <- "yes"
tcga_tml_pool2_idx %>%
  mutate(variant_hash = paste(gsub("chr","",Chromosome),Start_Position,End_Position,sep=";")) %>%
  select(variant_hash, pool2) %>%
  unique() -> ttmlpool2

# annotate variants as in pool 1, 2, or both
tcga_tml %>%
  left_join(ttmlpool1,by=c("variant_hash")) %>%
  left_join(ttmlpool2,by=c("variant_hash")) %>%
  mutate(pool=ifelse(pool1 %in% c("yes") & pool2 %in% c("yes"),"both",
                     ifelse(pool1 %in% c("yes"),"pool1","pool2"))) %>%
  select(-pool1,-pool2) -> tcga_tml2

# 'truth' rates
tcga_truth <- read_tsv("/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.wes_muts.gt_vf0.rate_tbl.txt")

```
```{r TML sampling function}
# sampleTCGAtml function: Explore TCGA concordance for a given panel size
# 1) subset tcga calls to those at or above specific VF threshold 
# 3) calculate muts/mb and correlate with WES muts/mb and return
sampleTCGAtml = function(tcgadf = tcga_tml2, # tcga variants data frame
                odir = "/mnt/hgfs/danhovelson/analysis/tmb/tml", # output directory
                vfthresh = 0, # min variant fraction threshold
                tcgacallsprefix = "/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.wes_muts.gt_vf", # file prefix for tcga VF-filtered calls
                truth = "/mnt/hgfs/danhovelson/analysis/tmb/20180525.tcga_pancx_coding_variants.wes_muts.gt_vf0.rate_tbl.txt", # truth (WES) rates
                panel_size_mb = 1.7, # panel size, in megabases
                coverage = 0, # min TCGA coverage
                iter = 0 # optional: current iteration, defaults to 0
) {
  # eliminate funky scientific notation problems for large-ish panel size
  options(scipen=999)
  
  # mkdir output directory if it doesn't exist
  mkdirs(odir)
  
  # define panel size & integer vf threshold
  psize = gsub("\\.","pt",as.character(round(panel_size_mb,digits=2)))
  vf = vfthresh*100
  
  # read in 'truth' mut counts/rates
  tcgatruth <- read_tsv(truth)
  
  # filter tcga_tml calls to those:
  # 1) above variant fraction threshold
  tcga_sub <- tcgadf %>%
    filter(VF > vfthresh & t_depth > coverage)
  
  ##### overall TML panel
  # calculate sampled ('in-panel') mut counts/rates
  submuts <- as.data.frame(table(tcga_sub$Tumor_Sample_Barcode)) %>%
    mutate(panel_mutmb = Freq/panel_size_mb,
          panel_size = panel_size_mb,
          iter_num = iter,
          vfthresh=vf,
          coverage=coverage,
          group="overall")
  colnames(submuts) <- c("Tumor_Sample_Barcode","panel_muts","panel_muts_mb","panel_size","iter_num","vfthresh","coverage","group")
  
  # merge with 'truth' set
  submutsmerge <- submuts %>%
    left_join(tcgatruth,by=c("Tumor_Sample_Barcode"))
  
  # parse tumor sample barcode -- retain patient id only
  submuts2 <- cSplit(submutsmerge,
                    splitCols = "Tumor_Sample_Barcode",
                    sep="-",
                    drop = FALSE)
  submuts2$bcr_patient_barcode = paste(submuts2$Tumor_Sample_Barcode_1, submuts2$Tumor_Sample_Barcode_2, submuts2$Tumor_Sample_Barcode_3, sep="-")
  submutsout <- submuts2[,c("bcr_patient_barcode","wes_mut_count","wes_mut_mb","panel_muts","panel_muts_mb","panel_size","iter_num","vfthresh","coverage","group")]

  
  ##### totals: by pool
  # calculate sampled ('in-panel') mut counts/rates
  submutsp <- as.data.frame(table(tcga_sub$Tumor_Sample_Barcode,tcga_sub$pool)) %>%
    mutate(panel_mutmb = Freq/panel_size_mb,
          panel_size = panel_size_mb,
          iter_num = iter,
          vfthresh=vf,
          coverage=coverage)
  colnames(submutsp) <- c("Tumor_Sample_Barcode","group","panel_muts","panel_muts_mb","panel_size","iter_num","vfthresh","coverage")
  
  # merge with 'truth' set
  submutsmergep <- submutsp %>%
    left_join(tcgatruth,by=c("Tumor_Sample_Barcode"))
  
  # parse tumor sample barcode -- retain patient id only
  submuts2p <- cSplit(submutsmergep,
                    splitCols = "Tumor_Sample_Barcode",
                    sep="-",
                    drop = FALSE)
  submuts2p$bcr_patient_barcode = paste(submuts2p$Tumor_Sample_Barcode_1, submuts2p$Tumor_Sample_Barcode_2, submuts2p$Tumor_Sample_Barcode_3, sep="-")
  submutsoutp <- submuts2p[,c("bcr_patient_barcode","wes_mut_count","wes_mut_mb","panel_muts","panel_muts_mb","panel_size","iter_num","vfthresh","coverage","group")]
  
  # cat overall and group totals
  submutsoutfin <- rbind(submutsout,submutsoutp)
  
  remove(tcgatruth)
  return(submutsoutfin)
}
```


```{r}
# overall plot
test <- sampleTCGAtml()
ggplot(test,aes(panel_muts_mb,wes_mut_mb,color=group)) + 
  theme_bw() +
  geom_point() + 
  facet_wrap(~group,scales="free") +
  scale_y_log10() +
  scale_x_log10()

# vf thresholds
vfthresholds <- c(seq(0,0.25,0.05),0.5)

# loop through list of panel_sizes, iterating pre-specified number of times
for (i in 1:length(panel_sizes)) {
  #i = 1
  for (j in 1:length(niterations)) {
    #j = 1
    for(k in 1:length(vfthresholds)) {
      #k = 1
      master_tbl = sampleTCGA(nlines=panel_sizes[i],iter=niterations[j],vfthresh=vfthresholds[k])
      vf <- vfthresholds[k]*100
      iternum <- niterations[j]
      psize = gsub("\\.","pt",as.character(round(master_tbl[1,"panel_size"],digits=2)))
      data_file = paste0("/mnt/hgfs/danhovelson/analysis/tmb/tmp/regions_",psize,"mb_n",iternum,"_vf",vf,".data_tbl.txt")
      write_tsv(master_tbl,path=data_file)
      remove(master_tbl)
    }
  }
}

```



```{r eval=FALSE}
# read in sample index
tcga_samps <- read_csv("/mnt/hgfs/danhovelson/analysis/tmb/tcga_sample_table.csv")

# combine exome/tcga muts (quartiles)
compcalls <- exomemuts %>%
  left_join(tmlmuts,by=c("Var1","Var2")) %>%
  mutate(cat = "deciles")
colnames(compcalls) = c("Tumor_Sample_Barcode","VFgroup","ExomeMuts","TMLMuts","Cat")

# combine exome/tcga muts (10% threshold)
compcalls1 <- exomemuts1 %>%
  left_join(tmlmuts1,by=c("Var1","Var2"))  %>%
  mutate(cat = "pct10")
colnames(compcalls1) = c("Tumor_Sample_Barcode","VFgroup","ExomeMuts","TMLMuts","Cat")

# combine exome/tcga muts (15% threshold)
compcalls2 <- exomemuts2 %>%
  left_join(tmlmuts2,by=c("Var1","Var2"))  %>%
  mutate(cat = "pct15")
colnames(compcalls2) = c("Tumor_Sample_Barcode","VFgroup","ExomeMuts","TMLMuts","Cat")

compcalls_ = rbind(compcalls,compcalls1,compcalls2)

compcallsout <- cSplit(compcalls_,
                    splitCols = "Tumor_Sample_Barcode",
                    sep="-",
                    drop = FALSE)

# parse tumor specimen barcode - retain sample barcode only
compcallsout$bcr_patient_barcode = paste(compcallsout$Tumor_Sample_Barcode_1,compcallsout$Tumor_Sample_Barcode_2,compcallsout$Tumor_Sample_Barcode_3,sep="-")
```

```{r fig1, fig.height = 11, fig.width = 9}
# Plot comparison to TCGA (all WES mutation calls)
# calculate muts/mb
compcallsfin <- compcallsout %>%
  select(bcr_patient_barcode,Tumor_Sample_Barcode,VFgroup,ExomeMuts,TMLMuts,Cat) %>%
  left_join(tcga_samps,by=c("bcr_patient_barcode")) %>%
  mutate(TMLMuts2 = ifelse(is.na(TMLMuts),0.1,TMLMuts)) %>%
  mutate(TMLmb = TMLMuts2/1.5,Exomemb=ExomeMuts/30)

# TCGA: mut count comparison
ggplot(compcallsfin[compcallsfin$TMLMuts2 > 0.1,],aes(Exomemb,TMLmb,color=type)) + 
  theme_bw() +
  geom_point(aes(color=type),shape=21,size=1) +
  geom_smooth(method="lm",color="red") +
  scale_y_log10("WES Coding Muts/Mb in TML Regions (by Variant Fraction)",breaks=c(0.1,1,10,100,1000),label=comma) +
  scale_x_log10("All WES Coding Muts/Mb",breaks=c(0.1,1,10,100,1000),label=comma) +
  facet_wrap(~VFgroup) +
  ggtitle("TCGA Variant Calls: WES vs TML Regions\n[n=9,561 TCGA Samples w/ 1+ Variant in TML regions]") +
  theme(legend.position="bottom") 
```

```{r fig2, fig.height = 5, fig.width = 7}
# Plot TCGA (VF >10% / >15%)
compcallsthresh <- compcallsfin[compcallsfin$VFgroup %in% c("(0.1,1]","(0.15,1]") & compcallsfin$ExomeMuts > 0 & !is.na(compcallsfin$type),]

# TCGA: mut count comparison
ggplot(compcallsthresh, aes(Exomemb,TMLmb,color=type)) + 
  theme_bw() +
  geom_point(aes(color=type),shape=21,size=1) +
  geom_smooth(method="lm",color="red") +
  scale_y_log10("WES Coding Muts/Mb in TML Regions (by Variant Fraction)",breaks=c(0.1,1,10,100,1000),label=comma) +
  scale_x_log10("All WES Coding Muts/Mb",breaks=c(0.1,1,10,100,1000),label=comma) +
  facet_wrap(~VFgroup) +
  ggtitle("TCGA Variant Calls: WES vs TML Regions\n[n=9,561 TCGA Samples w/ 1+ Variant in TML regions]") +
  theme(legend.position="bottom")

# correlations
cor(compcallsthresh[compcallsthresh$VFgroup %in% c("(0.1,1]"),]$Exomemb,
     compcallsthresh[compcallsthresh$VFgroup %in% c("(0.1,1]"),]$TMLmb)

cor(compcallsthresh[compcallsthresh$VFgroup %in% c("(0.15,1]"),]$Exomemb,
     compcallsthresh[compcallsthresh$VFgroup %in% c("(0.15,1]"),]$TMLmb)

```

```{r TML pool 1/2}
head(tcga_tml
     )
```






```{r sensitivity/specificity calculations}
# sens/spec calculations
vf0df %>%
  left_join(master_mut_sum,by=c("bcr_patient_barcode")) %>%
  mutate(rate_group_panel = cut(panel_rate,c(0,10,20,40,100,1000))) %>%
  filter(panel_size==5 & vfthresh==50) %>%
  group_by(panel_size,vfthresh) -> tmp

tgroups <- unique(tmp$rate_group)
for(i in tgroups {
  i = 1
  cgrp <- as.character(tgroups[i])
  t <- tmp[tmp$rate_group %in% cgrp,]
  t$ng <- if_else(t$rate_group %in% c(cgrp),0,1)
  t$npg <- if_else(t$rate_group_panel %in% c(cgrp),0,1)
  tbl <- as.data.frame(table(t$ng,t$npg))
  tp <- tbl[tbl$Var1 == 0 & tbl$Var2 == 0,]$Freq
  tn <- tbl[tbl$Var1 == 1]
  model = sensitivity(],x=ng,class=npg,                    method=maximize_metric,
                    metric=sum_sens_spec)
}

t <- as.data.frame(table(tmp$rate_group_panel,tmp$rate_group))
caret::sensitivity(table(tmp$rate_group_panel,tmp$rate_group),positive=rownames())


